File: elementary_datastructures/stack/Stack.java
public class Stack<T> {
    private int max_size, top;
    private T[] array;
    public Stack(int max_size){
        this.max_size = max_size;
        @SuppressWarnings("unchecked")
        T[] array = (T[])new Object[max_size];
        this.array = array;
        this.top = -1;
    }

    public boolean empty() {
        return this.top == -1;
    }

    public boolean full() {
        return this.top == this.max_size - 1;
    }

    public void push(T x) throws StackOverflowException{
        if (this.full())
            throw new StackOverflowException();
        this.top ++;
        this.array[top] = x;
    }

    public T pop() throws StackIsEmpty{
        if (this.empty())
            throw new StackIsEmpty();
        this.top --;
        return this.array[top+1];
    }
}

class StackOverflowException extends Exception {}
class StackIsEmpty extends Exception  {}

File: elementary_datastructures/stack/stack.cpp
// Implementation of stack data structure in C++
#include <bits/stdc++.h>
using namespace std;
//Size of stack:
#define SIZE 1000

template <typename T>

//Declare of stack class
class Stack{
    // top index holder.
    int top;
    public:
        T array[SIZE];
        Stack()
        {
            top = -1;
        }
        bool push(T x); //Append new number to stack.
        T pop(); // Pop last element of stack.
        T peek();// Return last element of stack.
        bool isEmpty(); // Check if stack is empty or not.
};

template <typename T>
bool Stack<T>::push(T x)
{
    if (top >= SIZE - 1)
    {
        cout << "ERROR! Stack overflow!" << endl;
        return false;
    }
    else
    {
        array[++top] = x;
        return true;
    }
}

template <typename T>
T Stack<T>::pop()
{
    if(top < 0)
    {
        cout << "ERROR! Stack underflow!" << endl;
        return (T)NULL;
    }
    else
    {
        T x = array[top--];
        return x;
    }
}

template <typename T>
T Stack<T>::peek()
{
    if (top < 0)
    {
        cout << "ERROR! Stack is empty!" << endl;
        return (T)NULL;
    }
    else
    {
        T x = array[top];
        return x;
    }
}

template <typename T>
bool Stack<T>::isEmpty()
{
    return (top<0);
}